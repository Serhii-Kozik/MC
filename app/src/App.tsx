import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import PageVisibility from 'react-page-visibility';

import { setAthletesData, setAuthleteDateType, Athlet } from './actions';
import { State } from './reducers';
import './App.scss';
import List from './components/List/List';

export interface Props {
  athlets: Athlet[];
  setAthletes: Function;
}

interface ComponentState {
  update: boolean;
  visible: boolean;
}

export class App extends Component<Props, ComponentState> {
// define socket
  socket: any;

  constructor(props: Props){
    super(props);

    this.state = {
      update: false,
      visible: true,
    }

  }

  componentDidMount(){
    this.openWebsocketConnection();
  }

  closeWebsocketConnection() {
    this.socket.close();
  }

  openWebsocketConnection(){
    this.socket = new WebSocket('ws://localhost:3355/');
    this.socket.onmessage = ({ data }: any) => {

        const response = JSON.parse(data);
        if (response.length === 0 ) { 
          setTimeout(() => {
            this.props.setAthletes([])
          }, 15000)};

        const validData = this.validateResponse(response);

        if (!validData) return;

        this.processData(validData);
    }
  }

  validateResponse(response: any){
    return response
      && Array.isArray(response)
      && response.length > 0
      && response[0].id
      && response[0].code
      && response[0].full_name
      && response[0].entering
      ? response
      : false
    ;
  }

  processData(data: any) {

    // if response contain more than 1 athlete 
    // (means new connection or gap in session) 
    // - we should update whole state 
    if (data.length > 1){
      this.props.setAthletes(data);
    } else {
      this.updateOrAddAthlet(data[0]);
    }
  }

  updateOrAddAthlet(update: Athlet){ 
    // check if recived athlet already in the list
    let isAthetInTheList: boolean | number = false;

    this.props.athlets.forEach((athlet, index) => {
      if (athlet.id === update.id) {
        isAthetInTheList = index;
      }
    });
    // if not add to the list
    if (isAthetInTheList === false) {

      let newList = this.props.athlets;
      newList.push(update);
      this.props.setAthletes(newList);
      this.setState({update: !this.state.update});

    } else {
      // if is in the list - update
      let updatedList = this.props.athlets;
      updatedList[isAthetInTheList] = update;
      this.props.setAthletes(updatedList);
      this.setState({update: !this.state.update});

    }
  }
    // close connection if browser window inactive and reconnect if active
    handleVisibilityChange = (isVisible: boolean) => {

      if (isVisible) {
          this.openWebsocketConnection();
      } else {
        this.closeWebsocketConnection();
      }
  }

  render(): any {

    return (
      <PageVisibility onChange={this.handleVisibilityChange}>
        <div className="app">
          <List athlets={this.props.athlets}/>
        </div>
      </PageVisibility>
    );
  }
}
const mapStateToProps = ({athlets}:State) => {
  return {athlets};
};

const mapDispatchToProps = (dispatch: Dispatch<setAuthleteDateType>) => {
  return {
      setAthletes: (data: State) =>
      dispatch(setAthletesData(data))

  };
} 
export default connect(mapStateToProps, mapDispatchToProps)(App);
