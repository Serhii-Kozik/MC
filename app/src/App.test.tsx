import React from 'react';
import { App, Props } from './App';
import Enzyme ,{ shallow }from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('APP component', () => {
  let component: any;
  let props: Props;

  beforeEach(() => {
    props = {
      athlets: [
        {
          id: 1,
          code: '123',
          start_nr: '123',
          full_name: 'first name',
          entering: 1,
          finish:2
        },
        {
          id: 2,
          code: '124',
          start_nr: '124',
          full_name: 'first name',
          entering: 2,
          finish:1
        },
      ],
      setAthletes: jest.fn(),
      
    }
    component = shallow(<App {...props}/>);
  });

  it('to match snapshot', () => {
    
    expect(component).toMatchSnapshot();
  });

  it('tryes to open websocket connection', () => {
    const spy = jest.spyOn(component.instance(), 'openWebsocketConnection');
    component.instance().componentDidMount();
    expect(spy).toBeCalled();
  });

  it('validates data correctly', () => {
    const validData = [
      {
        id: 2,
        code: '124',
        start_nr: '124',
        full_name: 'first name',
        entering: 2,
        finish:1
      }
    ];
    const valid = component.instance().validateResponse(validData);
    const invalidArray = component.instance().validateResponse([]);
    const invalidString = component.instance().validateResponse('string');

    expect(valid).toBe(validData);
    expect(invalidArray).toBe(false);
    expect(invalidString).toBe(false);
  });

  it('processData sets state if response array legth > 1', () => {
    component.instance().processData(props.athlets);
    expect(component.instance().props.setAthletes).toBeCalledWith(props.athlets);
  });

  it('processData calls update function if if response array legth = 1', () => {
    const responce = [
      {
        id: 1,
        code: '123',
        start_nr: '123',
        full_name: 'first name',
        entering: 1,
        finish:2
      },
    ];
    const spy = jest.spyOn(component.instance(), 'updateOrAddAthlet')
    component.instance().processData(responce);
    expect(spy).toBeCalledWith(responce[0]);
  });

  it('handleVisibilityChange sets correct state', () => {
    const spyOpen = jest.spyOn(component.instance(), 'openWebsocketConnection');
    const spyClose = jest.spyOn(component.instance(), 'closeWebsocketConnection');

    component.instance().handleVisibilityChange(true);
    expect(spyOpen).toBeCalled();

    component.instance().handleVisibilityChange(false);
    expect(spyClose).toBeCalled();
  });

  it('updateOrAddAthlet ubdates value if athlet in the list', () => {
    const spySetState = jest.spyOn(component.instance(), 'setState');
    component.instance().updateOrAddAthlet(props.athlets[0]);
    expect(component.instance().props.setAthletes).toBeCalledWith(props.athlets);
    expect(spySetState).toBeCalledWith({update: true});
    
  });

  it('updateOrAddAthlet adds new athlet if not in the list', () => {
    const spySetState = jest.spyOn(component.instance(), 'setState');
    const newAthlet = {
      id: 3,
      code: '124',
      start_nr: '124',
      full_name: 'first name',
      entering: 4,
      finish:5
    };
    const newList = props.athlets;
          newList.push(newAthlet);
    component.instance().updateOrAddAthlet(newAthlet);
    expect(component.instance().props.setAthletes).toBeCalledWith(newList);
    expect(spySetState).toBeCalledWith({update: true});
  });

});
