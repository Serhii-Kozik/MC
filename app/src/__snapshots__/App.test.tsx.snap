// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APP component to match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    athlets={
      Array [
        Object {
          "code": "123",
          "entering": 1,
          "finish": 2,
          "full_name": "first name",
          "id": 1,
          "start_nr": "123",
        },
        Object {
          "code": "124",
          "entering": 2,
          "finish": 1,
          "full_name": "first name",
          "id": 2,
          "start_nr": "124",
        },
      ]
    }
    setAthletes={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <div
        className="App"
      >
        <List
          athlets={
            Array [
              Object {
                "code": "123",
                "entering": 1,
                "finish": 2,
                "full_name": "first name",
                "id": 1,
                "start_nr": "123",
              },
              Object {
                "code": "124",
                "entering": 2,
                "finish": 1,
                "full_name": "first name",
                "id": 2,
                "start_nr": "124",
              },
            ]
          }
        />
      </div>,
      "onChange": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <List
          athlets={
            Array [
              Object {
                "code": "123",
                "entering": 1,
                "finish": 2,
                "full_name": "first name",
                "id": 1,
                "start_nr": "123",
              },
              Object {
                "code": "124",
                "entering": 2,
                "finish": 1,
                "full_name": "first name",
                "id": 2,
                "start_nr": "124",
              },
            ]
          }
        />,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "athlets": Array [
            Object {
              "code": "123",
              "entering": 1,
              "finish": 2,
              "full_name": "first name",
              "id": 1,
              "start_nr": "123",
            },
            Object {
              "code": "124",
              "entering": 2,
              "finish": 1,
              "full_name": "first name",
              "id": 2,
              "start_nr": "124",
            },
          ],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          className="App"
        >
          <List
            athlets={
              Array [
                Object {
                  "code": "123",
                  "entering": 1,
                  "finish": 2,
                  "full_name": "first name",
                  "id": 1,
                  "start_nr": "123",
                },
                Object {
                  "code": "124",
                  "entering": 2,
                  "finish": 1,
                  "full_name": "first name",
                  "id": 2,
                  "start_nr": "124",
                },
              ]
            }
          />
        </div>,
        "onChange": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <List
            athlets={
              Array [
                Object {
                  "code": "123",
                  "entering": 1,
                  "finish": 2,
                  "full_name": "first name",
                  "id": 1,
                  "start_nr": "123",
                },
                Object {
                  "code": "124",
                  "entering": 2,
                  "finish": 1,
                  "full_name": "first name",
                  "id": 2,
                  "start_nr": "124",
                },
              ]
            }
          />,
          "className": "App",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "athlets": Array [
              Object {
                "code": "123",
                "entering": 1,
                "finish": 2,
                "full_name": "first name",
                "id": 1,
                "start_nr": "123",
              },
              Object {
                "code": "124",
                "entering": 2,
                "finish": 1,
                "full_name": "first name",
                "id": 2,
                "start_nr": "124",
              },
            ],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
